node-pty doesn't build on Node 12 (V8 7.4)

## Environment details

- OS: macOS
- OS version: 10.14.3
- node-pty version: 0.8.1

## Issue description

In node 10, we saw deprecation notices during the build but node-pty still worked.

As of today's release of Node 12, node-pty no longer builds.

I will take a stab at an update and submit a PR if I have any luck.

```
21:29 $ npm i node-pty@0.8.1

> node-pty@0.8.1 install /Users/mogden/repos/github.com/machellerogden/node-pty-test/node_modules/node-pty
> node scripts/install.js

  CXX(target) Release/obj.target/pty/src/unix/pty.o
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:222:
In file included from ../../nan/nan_converters.h:67:
../../nan/nan_converters_43_inl.h:22:1: warning: 'ToBoolean' is deprecated: ToBoolean can never throw. Use Local version. [-Wdeprecated-declarations]
X(Boolean)
^
../../nan/nan_converters_43_inl.h:18:12: note: expanded from macro 'X'
      val->To ## TYPE(isolate->GetCurrentContext())                            \
           ^
<scratch space>:23:1: note: expanded from here
ToBoolean
^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2523:3: note: 'ToBoolean' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("ToBoolean can never throw. Use Local version.",
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:222:
In file included from ../../nan/nan_converters.h:67:
../../nan/nan_converters_43_inl.h:40:1: warning: 'BooleanValue' is deprecated: BooleanValue can never throw. Use Isolate version. [-Wdeprecated-declarations]
X(bool, Boolean)
^
../../nan/nan_converters_43_inl.h:37:15: note: expanded from macro 'X'
  return val->NAME ## Value(isolate->GetCurrentContext());                     \
              ^
<scratch space>:30:1: note: expanded from here
BooleanValue
^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2561:3: note: 'BooleanValue' has been explicitly marked deprecated here
  V8_DEPRECATED("BooleanValue can never throw. Use Isolate version.",
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:307:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:223:
In file included from ../../nan/nan_new.h:189:
../../nan/nan_implementation_12_inl.h:356:37: error: too few arguments to function call, expected 2, have 1
  return v8::StringObject::New(value).As<v8::StringObject>();
         ~~~~~~~~~~~~~~~~~~~~~      ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:5380:3: note: 'New' declared here
  static Local<Value> New(Isolate* isolate, Local<String> value);
  ^
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:223:
In file included from ../../nan/nan_new.h:189:
../../nan/nan_implementation_12_inl.h:356:58: error: expected '(' for function-style cast or type construction
  return v8::StringObject::New(value).As<v8::StringObject>();
                                         ~~~~~~~~~~~~~~~~^
../../nan/nan_implementation_12_inl.h:356:60: error: expected expression
  return v8::StringObject::New(value).As<v8::StringObject>();
                                                           ^
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:2722:
../../nan/nan_object_wrap.h:24:25: error: no member named 'IsNearDeath' in 'Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    assert(persistent().IsNearDeath());
           ~~~~~~~~~~~~ ^
/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
In file included from ../src/unix/pty.cc:20:
In file included from ../../nan/nan.h:2722:
../../nan/nan_object_wrap.h:127:26: error: no member named 'IsNearDeath' in 'Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >'
    assert(wrap->handle_.IsNearDeath());
           ~~~~~~~~~~~~~ ^
/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
../src/unix/pty.cc:154:39: error: no matching member function for call to 'ToString'
  v8::String::Utf8Value file(info[0]->ToString());
                             ~~~~~~~~~^~~~~~~~
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2528:44: note: candidate function not viable: requires single argument 'context', but no arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<String> ToString(
                                           ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2544:35: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
                    Local<String> ToString(Isolate* isolate) const);
                                  ^
../src/unix/pty.cc:165:69: error: no matching member function for call to 'ToString'
    v8::String::Utf8Value arg(argv_->Get(Nan::New<v8::Integer>(i))->ToString());
                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2528:44: note: candidate function not viable: requires single argument 'context', but no arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<String> ToString(
                                           ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2544:35: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
                    Local<String> ToString(Isolate* isolate) const);
                                  ^
../src/unix/pty.cc:165:38: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
    v8::String::Utf8Value arg(argv_->Get(Nan::New<v8::Integer>(i))->ToString());
                                     ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/unix/pty.cc:176:69: error: no matching member function for call to 'ToString'
    v8::String::Utf8Value pair(env_->Get(Nan::New<v8::Integer>(i))->ToString());
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2528:44: note: candidate function not viable: requires single argument 'context', but no arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<String> ToString(
                                           ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2544:35: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
                    Local<String> ToString(Isolate* isolate) const);
                                  ^
../src/unix/pty.cc:176:38: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
    v8::String::Utf8Value pair(env_->Get(Nan::New<v8::Integer>(i))->ToString());
                                     ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/unix/pty.cc:181:39: error: no matching member function for call to 'ToString'
  v8::String::Utf8Value cwd_(info[3]->ToString());
                             ~~~~~~~~~^~~~~~~~
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2528:44: note: candidate function not viable: requires single argument 'context', but no arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<String> ToString(
                                           ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2544:35: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
                    Local<String> ToString(Isolate* isolate) const);
                                  ^
../src/unix/pty.cc:186:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_col = info[4]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:187:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_row = info[5]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:195:16: error: no matching member function for call to 'ToBoolean'
  if (info[8]->ToBoolean()->Value()) {
      ~~~~~~~~~^~~~~~~~~
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2524:63: note: candidate function not viable: requires single argument 'context', but no arguments were provided
                    V8_WARN_UNUSED_RESULT MaybeLocal<Boolean> ToBoolean(
                                                              ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2540:18: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
  Local<Boolean> ToBoolean(Isolate* isolate) const;
                 ^
../src/unix/pty.cc:230:35: error: too few arguments to function call, single argument 'context' was not specified
  int uid = info[6]->IntegerValue();
            ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:231:35: error: too few arguments to function call, single argument 'context' was not specified
  int gid = info[7]->IntegerValue();
            ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:315:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_col = info[0]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:316:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_row = info[1]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:360:34: error: too few arguments to function call, single argument 'context' was not specified
  int fd = info[0]->IntegerValue();
           ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:363:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_col = info[1]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/unix/pty.cc:364:39: error: too few arguments to function call, single argument 'context' was not specified
  winp.ws_row = info[2]->IntegerValue();
                ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8.h:2565:3: note: 'IntegerValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<int64_t> IntegerValue(
  ^
/Users/mogden/.node-gyp/12.0.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
4 warnings and 20 errors generated.
make: *** [Release/obj.target/pty/src/unix/pty.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/mogden/n/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:256:12)
gyp ERR! System Darwin 18.2.0
gyp ERR! command "/Users/mogden/n/bin/node" "/Users/mogden/n/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/mogden/repos/github.com/machellerogden/node-pty-test/node_modules/node-pty
gyp ERR! node -v v12.0.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm WARN node-pty-test@1.0.0 No description
npm WARN node-pty-test@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-pty@0.8.1 install: `node scripts/install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-pty@0.8.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mogden/.npm/_logs/2019-04-24T02_29_19_741Z-debug.log
```

thanks @Tyriar, I will see what I can do. We are using node-pty on NWJS and it is much further than Electron. Already on Chrome 74 and Node 12.

I also noticed:

> ..\src\win\conpty.cc(249): error C2660: 'v8::Function::Call': function does not take 3 arguments [xxx\node_modules\node-pty\build\conpty.vcxproj]

Thanks for you great work btw - it helped us a lot to get a full functional terminal on Windows.

--- a/package.json
+++ b/package.json
@@ -42,7 +42,7 @@
     "prepublish": "npm run tsc"
   },
   "dependencies": {
-    "nan": "2.12.1"
+    "nan": "^2.13.2"
   },
   "devDependencies": {
     "@types/mocha": "^5.0.0",

--- a/src/unix/pty.cc
+++ b/src/unix/pty.cc
@@ -192,7 +192,7 @@ NAN_METHOD(PtyFork) {
   struct termios t = termios();
   struct termios *term = &t;
   term->c_iflag = ICRNL | IXON | IXANY | IMAXBEL | BRKINT;
-  if (info[8]->BooleanValue(info.GetIsolate())) {
+  if (info[8]->BooleanValue(Nan::GetCurrentContext()).FromJust()) {
 #if defined(IUTF8)
     term->c_iflag |= IUTF8;
 #endif

--- a/src/unix/pty.cc
+++ b/src/unix/pty.cc
@@ -151,7 +151,7 @@ NAN_METHOD(PtyFork) {
   signal(SIGINT, SIG_DFL);
 
   // file
-  v8::String::Utf8Value file(info[0]->ToString());
+  Nan::Utf8String file(info[0]);
 
   // args
   int i = 0;
@@ -162,7 +162,7 @@ NAN_METHOD(PtyFork) {
   argv[0] = strdup(*file);
   argv[argl-1] = NULL;
   for (; i < argc; i++) {
-    v8::String::Utf8Value arg(argv_->Get(Nan::New<v8::Integer>(i))->ToString());
+    Nan::Utf8String arg(Nan::Get(argv_, i).ToLocalChecked());
     argv[i+1] = strdup(*arg);
   }
 
@@ -173,18 +173,18 @@ NAN_METHOD(PtyFork) {
   char **env = new char*[envc+1];
   env[envc] = NULL;
   for (; i < envc; i++) {
-    v8::String::Utf8Value pair(env_->Get(Nan::New<v8::Integer>(i))->ToString());
+    Nan::Utf8String pair(Nan::Get(env_, i).ToLocalChecked());
     env[i] = strdup(*pair);
   }
 
   // cwd
-  v8::String::Utf8Value cwd_(info[3]->ToString());
+  Nan::Utf8String cwd_(info[3]);
   char *cwd = strdup(*cwd_);
 
   // size
   struct winsize winp;
-  winp.ws_col = info[4]->IntegerValue();
-  winp.ws_row = info[5]->IntegerValue();
+  winp.ws_col = info[4]->IntegerValue(Nan::GetCurrentContext()).FromJust();
+  winp.ws_row = info[5]->IntegerValue(Nan::GetCurrentContext()).FromJust();
   winp.ws_xpixel = 0;
   winp.ws_ypixel = 0;
 
@@ -192,7 +192,7 @@ NAN_METHOD(PtyFork) {
   struct termios t = termios();
   struct termios *term = &t;
   term->c_iflag = ICRNL | IXON | IXANY | IMAXBEL | BRKINT;
-  if (info[8]->ToBoolean()->Value()) {
+  if (info[8]->BooleanValue(info.GetIsolate())) {
 #if defined(IUTF8)
     term->c_iflag |= IUTF8;
 #endif
@@ -227,8 +227,8 @@ NAN_METHOD(PtyFork) {
   cfsetospeed(term, B38400);
 
   // uid / gid
-  int uid = info[6]->IntegerValue();
-  int gid = info[7]->IntegerValue();
+  int uid = info[6]->IntegerValue(Nan::GetCurrentContext()).FromJust();
+  int gid = info[7]->IntegerValue(Nan::GetCurrentContext()).FromJust();
 
   // fork the pty
   int master = -1;
@@ -312,8 +312,8 @@ NAN_METHOD(PtyOpen) {
 
   // size
   struct winsize winp;
-  winp.ws_col = info[0]->IntegerValue();
-  winp.ws_row = info[1]->IntegerValue();
+  winp.ws_col = info[0]->IntegerValue(Nan::GetCurrentContext()).FromJust();
+  winp.ws_row = info[1]->IntegerValue(Nan::GetCurrentContext()).FromJust();
   winp.ws_xpixel = 0;
   winp.ws_ypixel = 0;
 
@@ -357,11 +357,11 @@ NAN_METHOD(PtyResize) {
     return Nan::ThrowError("Usage: pty.resize(fd, cols, rows)");
   }
 
-  int fd = info[0]->IntegerValue();
+  int fd = info[0]->IntegerValue(Nan::GetCurrentContext()).FromJust();
 
   struct winsize winp;
-  winp.ws_col = info[1]->IntegerValue();
-  winp.ws_row = info[2]->IntegerValue();
+  winp.ws_col = info[1]->IntegerValue(Nan::GetCurrentContext()).FromJust();
+  winp.ws_row = info[2]->IntegerValue(Nan::GetCurrentContext()).FromJust();
   winp.ws_xpixel = 0;
   winp.ws_ypixel = 0;
 
@@ -384,9 +384,9 @@ NAN_METHOD(PtyGetProc) {
     return Nan::ThrowError("Usage: pty.process(fd, tty)");
   }
 
-  int fd = info[0]->IntegerValue();
+  int fd = info[0]->IntegerValue(Nan::GetCurrentContext()).FromJust();
 
-  v8::String::Utf8Value tty_(info[1]->ToString());
+  Nan::Utf8String tty_(info[1]);
   char *tty = strdup(*tty_);
   char *name = pty_getproc(fd, tty);
   free(tty);
@@ -700,18 +700,10 @@ pty_forkpty(int *amaster,
 
 NAN_MODULE_INIT(init) {
   Nan::HandleScope scope;
-  Nan::Set(target,
-           Nan::New<v8::String>("fork").ToLocalChecked(),
-           Nan::New<v8::FunctionTemplate>(PtyFork)->GetFunction());
-  Nan::Set(target,
-           Nan::New<v8::String>("open").ToLocalChecked(),
-           Nan::New<v8::FunctionTemplate>(PtyOpen)->GetFunction());
-  Nan::Set(target,
-           Nan::New<v8::String>("resize").ToLocalChecked(),
-           Nan::New<v8::FunctionTemplate>(PtyResize)->GetFunction());
-  Nan::Set(target,
-           Nan::New<v8::String>("process").ToLocalChecked(),
-           Nan::New<v8::FunctionTemplate>(PtyGetProc)->GetFunction());
+  Nan::Export(target, "fork", PtyFork);
+  Nan::Export(target, "open", PtyOpen);
+  Nan::Export(target, "resize", PtyResize);
+  Nan::Export(target, "process", PtyGetProc);
 }
 
 NODE_MODULE(pty, init)

--- a/src/unixTerminal.ts
+++ b/src/unixTerminal.ts
@@ -3,7 +3,6 @@
  * Copyright (c) 2016, Daniel Imms (MIT License).
  * Copyright (c) 2018, Microsoft Corporation (MIT License).
  */
-
 import * as net from 'net';
 import { Terminal, DEFAULT_COLS, DEFAULT_ROWS } from './terminal';
 import { IProcessEnv, IPtyForkOptions, IPtyOpenOptions } from './interfaces';
@@ -276,11 +275,10 @@ export class UnixTerminal extends Terminal {
  */
 class PipeSocket extends net.Socket {
   constructor(fd: number) {
-    const tty = (<any>process).binding('tty_wrap');
-    const guessHandleType = tty.guessHandleType;
-    tty.guessHandleType = () => 'PIPE';
+    const { Pipe, constants } = (<any>process).binding('pipe_wrap')
     // @types/node has fd as string? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18275
-    super({ fd: <any>fd });
-    tty.guessHandleType = guessHandleType;
+    const handle = new Pipe(constants.SOCKET);
+    handle.open(fd);
+    super(<any>{ handle });
   }
 }

--- a/src/unixTerminal.ts
+++ b/src/unixTerminal.ts
@@ -275,7 +275,7 @@ export class UnixTerminal extends Terminal {
  */
 class PipeSocket extends net.Socket {
   constructor(fd: number) {
-    const { Pipe, constants } = (<any>process).binding('pipe_wrap')
+    const { Pipe, constants } = (<any>process).binding('pipe_wrap'); // tslint:disable-line
     // @types/node has fd as string? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18275
     const handle = new Pipe(constants.SOCKET);
     handle.open(fd);

--- a/src/win/conpty.cc
+++ b/src/win/conpty.cc
@@ -40,7 +40,7 @@ struct pty_baton {
 
   HANDLE hShell;
   HANDLE hWait;
-  Nan::Persistent<v8::Function> cb;
+  Nan::Callback cb;
   uv_async_t async;
   uv_thread_t tid;
 
@@ -204,20 +204,20 @@ static NAN_METHOD(PtyStartProcess) {
   if (SUCCEEDED(hr)) {
     // We were able to instantiate a conpty
     const int ptyId = InterlockedIncrement(&ptyCounter);
-    marshal->Set(Nan::New<v8::String>("pty").ToLocalChecked(), Nan::New<v8::Number>(ptyId));
+    Nan::Set(marshal, Nan::New<v8::String>("pty").ToLocalChecked(), Nan::New<v8::Number>(ptyId));
     ptyHandles.insert(ptyHandles.end(), new pty_baton(ptyId, hIn, hOut, hpc));
   } else {
     Nan::ThrowError("Cannot launch conpty");
     return;
   }
 
-  marshal->Set(Nan::New<v8::String>("fd").ToLocalChecked(), Nan::New<v8::Number>(-1));
+  Nan::Set(marshal, Nan::New<v8::String>("fd").ToLocalChecked(), Nan::New<v8::Number>(-1));
   {
     std::string coninPipeNameStr(inName.begin(), inName.end());
-    marshal->Set(Nan::New<v8::String>("conin").ToLocalChecked(), Nan::New<v8::String>(coninPipeNameStr).ToLocalChecked());
+    Nan::Set(marshal, Nan::New<v8::String>("conin").ToLocalChecked(), Nan::New<v8::String>(coninPipeNameStr).ToLocalChecked());
 
     std::string conoutPipeNameStr(outName.begin(), outName.end());
-    marshal->Set(Nan::New<v8::String>("conout").ToLocalChecked(), Nan::New<v8::String>(conoutPipeNameStr).ToLocalChecked());
+    Nan::Set(marshal, Nan::New<v8::String>("conout").ToLocalChecked(), Nan::New<v8::String>(conoutPipeNameStr).ToLocalChecked());
   }
   info.GetReturnValue().Set(marshal);
 }
@@ -245,8 +245,9 @@ static void OnProcessExit(uv_async_t *async) {
   v8::Local<v8::Value> args[1] = {
     Nan::New<v8::Number>(exitCode)
   };
-  v8::Local<v8::Function> local = Nan::New(baton->cb);
-  local->Call(Nan::GetCurrentContext(), Nan::Null(), 1, args);
+
+  Nan::AsyncResource asyncResource("node-pty.callback");
+  baton->cb.Call(1, args, &asyncResource);
   // Clean up
   baton->cb.Reset();
 }
@@ -290,7 +291,7 @@ static NAN_METHOD(PtyConnect) {
   if (!envValues.IsEmpty()) {
     std::wstringstream envBlock;
     for(uint32_t i = 0; i < envValues->Length(); i++) {
-      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(envValues->Get(i))));
+      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(Nan::Get(envValues, i).ToLocalChecked())));
       envBlock << envValue << L'\0';
     }
     envBlock << L'\0';
@@ -308,12 +309,12 @@ static NAN_METHOD(PtyConnect) {
   // Attach the pseudoconsole to the client application we're creating
   STARTUPINFOEXW siEx{0};
   siEx.StartupInfo.cb = sizeof(STARTUPINFOEXW);
-  PSIZE_T size;
-  InitializeProcThreadAttributeList(NULL, 1, 0, size);
-  BYTE *attrList = new BYTE[*size];
+  SIZE_T size = 0;
+  InitializeProcThreadAttributeList(NULL, 1, 0, &size);
+  BYTE *attrList = new BYTE[size];
   siEx.lpAttributeList = reinterpret_cast<PPROC_THREAD_ATTRIBUTE_LIST>(attrList);
 
-  fSuccess = InitializeProcThreadAttributeList(siEx.lpAttributeList, 1, 0, size);
+  fSuccess = InitializeProcThreadAttributeList(siEx.lpAttributeList, 1, 0, &size);
   if (!fSuccess) {
     return throwNanError(&info, "InitializeProcThreadAttributeList failed", true);
   }
@@ -358,7 +359,7 @@ static NAN_METHOD(PtyConnect) {
 
   // Return
   v8::Local<v8::Object> marshal = Nan::New<v8::Object>();
-  marshal->Set(Nan::New<v8::String>("pid").ToLocalChecked(), Nan::New<v8::Number>(piClient.dwProcessId));
+  Nan::Set(marshal, Nan::New<v8::String>("pid").ToLocalChecked(), Nan::New<v8::Number>(piClient.dwProcessId));
   info.GetReturnValue().Set(marshal);
 }
 

--- a/src/win/conpty.cc
+++ b/src/win/conpty.cc
@@ -18,7 +18,7 @@
 #include <strsafe.h>
 #include "path_util.h"
 
-extern "C" void init(v8::Handle<v8::Object>);
+extern "C" void init(v8::Local<v8::Object>);
 
 // Taken from the RS5 Windows SDK, but redefined here in case we're targeting <= 17134
 #ifndef PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE
@@ -170,11 +170,11 @@ static NAN_METHOD(PtyStartProcess) {
     return;
   }
 
-  const std::wstring filename(path_util::to_wstring(v8::String::Utf8Value(info[0]->ToString())));
-  const SHORT cols = info[1]->Uint32Value();
-  const SHORT rows = info[2]->Uint32Value();
-  const bool debug = info[3]->ToBoolean()->IsTrue();
-  const std::wstring pipeName(path_util::to_wstring(v8::String::Utf8Value(info[4]->ToString())));
+  const std::wstring filename(path_util::to_wstring(Nan::Utf8String(info[0])));
+  const SHORT cols = info[1]->Uint32Value(Nan::GetCurrentContext()).FromJust();
+  const SHORT rows = info[2]->Uint32Value(Nan::GetCurrentContext()).FromJust();
+  const bool debug = info[3]->ToBoolean(Nan::GetCurrentContext()).ToLocalChecked()->IsTrue();
+  const std::wstring pipeName(path_util::to_wstring(Nan::Utf8String(info[4])));
 
   // use environment 'Path' variable to determine location of
   // the relative path that we have recieved (e.g cmd.exe)
@@ -242,12 +242,11 @@ static void OnProcessExit(uv_async_t *async) {
   GetExitCodeProcess(baton->hShell, &exitCode);
 
   // Call function
-  v8::Handle<v8::Value> args[1] = {
+  v8::Local<v8::Value> args[1] = {
     Nan::New<v8::Number>(exitCode)
   };
-  v8::Handle<v8::Function> local = Nan::New(baton->cb);
-  local->Call(Nan::GetCurrentContext()->Global(), 1, args);
-
+  v8::Local<v8::Function> local = Nan::New(baton->cb);
+  local->Call(Nan::GetCurrentContext(), Nan::Null(), 1, args);
   // Clean up
   baton->cb.Reset();
 }
@@ -272,10 +271,10 @@ static NAN_METHOD(PtyConnect) {
     return;
   }
 
-  const int id = info[0]->Int32Value();
-  const std::wstring cmdline(path_util::to_wstring(v8::String::Utf8Value(info[1]->ToString())));
-  const std::wstring cwd(path_util::to_wstring(v8::String::Utf8Value(info[2]->ToString())));
-  const v8::Handle<v8::Array> envValues = info[3].As<v8::Array>();
+  const int id = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  const std::wstring cmdline(path_util::to_wstring(Nan::Utf8String(info[1])));
+  const std::wstring cwd(path_util::to_wstring(Nan::Utf8String(info[2])));
+  const v8::Local<v8::Array> envValues = info[3].As<v8::Array>();
   const v8::Local<v8::Function> exitCallback = v8::Local<v8::Function>::Cast(info[4]);
 
   // Prepare command line
@@ -291,7 +290,7 @@ static NAN_METHOD(PtyConnect) {
   if (!envValues.IsEmpty()) {
     std::wstringstream envBlock;
     for(uint32_t i = 0; i < envValues->Length(); i++) {
-      std::wstring envValue(path_util::to_wstring(v8::String::Utf8Value(envValues->Get(i)->ToString())));
+      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(envValues->Get(i))));
       envBlock << envValue << L'\0';
     }
     envBlock << L'\0';
@@ -374,9 +373,9 @@ static NAN_METHOD(PtyResize) {
     return;
   }
 
-  int id = info[0]->Int32Value();
-  SHORT cols = info[1]->Uint32Value();
-  SHORT rows = info[2]->Uint32Value();
+  int id = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  SHORT cols = info[1]->Uint32Value(Nan::GetCurrentContext()).FromJust();
+  SHORT rows = info[2]->Uint32Value(Nan::GetCurrentContext()).FromJust();
 
   const pty_baton* handle = get_pty_baton(id);
 
@@ -404,7 +403,7 @@ static NAN_METHOD(PtyKill) {
     return;
   }
 
-  int id = info[0]->Int32Value();
+  int id = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
 
   const pty_baton* handle = get_pty_baton(id);
 
@@ -428,7 +427,7 @@ static NAN_METHOD(PtyKill) {
 * Init
 */
 
-extern "C" void init(v8::Handle<v8::Object> target) {
+extern "C" void init(v8::Local<v8::Object> target) {
   Nan::HandleScope scope;
   Nan::SetMethod(target, "startProcess", PtyStartProcess);
   Nan::SetMethod(target, "connect", PtyConnect);

--- a/src/win/conpty_console_list.cc
+++ b/src/win/conpty_console_list.cc
@@ -12,7 +12,7 @@ static NAN_METHOD(ApiConsoleProcessList) {
     return;
   }
 
-  const SHORT pid = info[0]->Uint32Value();
+  const SHORT pid = info[0]->Uint32Value(Nan::GetCurrentContext()).FromJust();
 
   if (!FreeConsole()) {
     Nan::ThrowError("FreeConsole failed");
@@ -35,7 +35,7 @@ static NAN_METHOD(ApiConsoleProcessList) {
   info.GetReturnValue().Set(result);
 }
 
-extern "C" void init(v8::Handle<v8::Object> target) {
+extern "C" void init(v8::Local<v8::Object> target) {
   Nan::HandleScope scope;
   Nan::SetMethod(target, "getConsoleProcessList", ApiConsoleProcessList);
 };

--- a/src/win/conpty_console_list.cc
+++ b/src/win/conpty_console_list.cc
@@ -30,7 +30,7 @@ static NAN_METHOD(ApiConsoleProcessList) {
 
   v8::Local<v8::Array> result = Nan::New<v8::Array>();
   for (DWORD i = 0; i < processCount; i++) {
-    result->Set(i, Nan::New<v8::Number>(processList[i]));
+    Nan::Set(result, i, Nan::New<v8::Number>(processList[i]));
   }
   info.GetReturnValue().Set(result);
 }

--- a/src/win/path_util.cc
+++ b/src/win/path_util.cc
@@ -11,7 +11,7 @@
 
 namespace path_util {
 
-const wchar_t* to_wstring(const v8::String::Utf8Value& str) {
+const wchar_t* to_wstring(const Nan::Utf8String& str) {
   const char *bytes = *str;
   unsigned int sizeOfStr = MultiByteToWideChar(CP_UTF8, 0, bytes, -1, NULL, 0);
   wchar_t *output = new wchar_t[sizeOfStr];

--- a/src/win/path_util.h
+++ b/src/win/path_util.h
@@ -13,7 +13,7 @@
 
 namespace path_util {
 
-const wchar_t* to_wstring(const v8::String::Utf8Value& str);
+const wchar_t* to_wstring(const Nan::Utf8String& str);
 bool file_exists(std::wstring filename);
 std::wstring get_shell_path(std::wstring filename);
 

--- a/src/win/winpty.cc
+++ b/src/win/winpty.cc
@@ -141,7 +141,7 @@ static NAN_METHOD(PtyStartProcess) {
     std::wstringstream envBlock;
 
     for(uint32_t i = 0; i < envValues->Length(); i++) {
-      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(envValues->Get(i))));
+      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(Nan::Get(envValues, i).ToLocalChecked())));
       envBlock << envValue << L'\0';
     }
 
@@ -216,20 +216,20 @@ static NAN_METHOD(PtyStartProcess) {
 
   // Set return values
   v8::Local<v8::Object> marshal = Nan::New<v8::Object>();
-  marshal->Set(Nan::New<v8::String>("innerPid").ToLocalChecked(), Nan::New<v8::Number>((int)GetProcessId(handle)));
-  marshal->Set(Nan::New<v8::String>("innerPidHandle").ToLocalChecked(), Nan::New<v8::Number>((int)handle));
-  marshal->Set(Nan::New<v8::String>("pid").ToLocalChecked(), Nan::New<v8::Number>((int)winpty_agent_process(pc)));
-  marshal->Set(Nan::New<v8::String>("pty").ToLocalChecked(), Nan::New<v8::Number>(InterlockedIncrement(&ptyCounter)));
-  marshal->Set(Nan::New<v8::String>("fd").ToLocalChecked(), Nan::New<v8::Number>(-1));
+  Nan::Set(marshal, Nan::New<v8::String>("innerPid").ToLocalChecked(), Nan::New<v8::Number>((int)GetProcessId(handle)));
+  Nan::Set(marshal, Nan::New<v8::String>("innerPidHandle").ToLocalChecked(), Nan::New<v8::Number>((int)handle));
+  Nan::Set(marshal, Nan::New<v8::String>("pid").ToLocalChecked(), Nan::New<v8::Number>((int)winpty_agent_process(pc)));
+  Nan::Set(marshal, Nan::New<v8::String>("pty").ToLocalChecked(), Nan::New<v8::Number>(InterlockedIncrement(&ptyCounter)));
+  Nan::Set(marshal, Nan::New<v8::String>("fd").ToLocalChecked(), Nan::New<v8::Number>(-1));
   {
     LPCWSTR coninPipeName = winpty_conin_name(pc);
     std::wstring coninPipeNameWStr(coninPipeName);
     std::string coninPipeNameStr(coninPipeNameWStr.begin(), coninPipeNameWStr.end());
-    marshal->Set(Nan::New<v8::String>("conin").ToLocalChecked(), Nan::New<v8::String>(coninPipeNameStr).ToLocalChecked());
+    Nan::Set(marshal, Nan::New<v8::String>("conin").ToLocalChecked(), Nan::New<v8::String>(coninPipeNameStr).ToLocalChecked());
     LPCWSTR conoutPipeName = winpty_conout_name(pc);
     std::wstring conoutPipeNameWStr(conoutPipeName);
     std::string conoutPipeNameStr(conoutPipeNameWStr.begin(), conoutPipeNameWStr.end());
-    marshal->Set(Nan::New<v8::String>("conout").ToLocalChecked(), Nan::New<v8::String>(conoutPipeNameStr).ToLocalChecked());
+    Nan::Set(marshal, Nan::New<v8::String>("conout").ToLocalChecked(), Nan::New<v8::String>(conoutPipeNameStr).ToLocalChecked());
   }
   info.GetReturnValue().Set(marshal);
 

--- a/src/win/winpty.cc
+++ b/src/win/winpty.cc
@@ -23,7 +23,7 @@
 /**
 * Misc
 */
-extern "C" void init(v8::Handle<v8::Object>);
+extern "C" void init(v8::Local<v8::Object>);
 
 #define WINPTY_DBG_VARIABLE TEXT("WINPTYDBG")
 
@@ -80,7 +80,7 @@ static NAN_METHOD(PtyGetExitCode) {
   }
 
   DWORD exitCode = 0;
-  GetExitCodeProcess((HANDLE)info[0]->IntegerValue(), &exitCode);
+  GetExitCodeProcess((HANDLE)info[0]->IntegerValue(Nan::GetCurrentContext()).FromJust(), &exitCode);
 
   info.GetReturnValue().Set(Nan::New<v8::Number>(exitCode));
 }
@@ -94,7 +94,7 @@ static NAN_METHOD(PtyGetProcessList) {
     return;
   }
 
-  int pid = info[0]->Int32Value();
+  int pid = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
 
   winpty_t *pc = get_pipe_handle(pid);
   if (pc == nullptr) {
@@ -129,19 +129,19 @@ static NAN_METHOD(PtyStartProcess) {
 
   std::stringstream why;
 
-  const wchar_t *filename = path_util::to_wstring(v8::String::Utf8Value(info[0]->ToString()));
-  const wchar_t *cmdline = path_util::to_wstring(v8::String::Utf8Value(info[1]->ToString()));
-  const wchar_t *cwd = path_util::to_wstring(v8::String::Utf8Value(info[3]->ToString()));
+  const wchar_t *filename = path_util::to_wstring(Nan::Utf8String(info[0]));
+  const wchar_t *cmdline = path_util::to_wstring(Nan::Utf8String(info[1]));
+  const wchar_t *cwd = path_util::to_wstring(Nan::Utf8String(info[3]));
 
   // create environment block
   std::wstring env;
-  const v8::Handle<v8::Array> envValues = v8::Handle<v8::Array>::Cast(info[2]);
+  const v8::Local<v8::Array> envValues = v8::Local<v8::Array>::Cast(info[2]);
   if (!envValues.IsEmpty()) {
 
     std::wstringstream envBlock;
 
     for(uint32_t i = 0; i < envValues->Length(); i++) {
-      std::wstring envValue(path_util::to_wstring(v8::String::Utf8Value(envValues->Get(i)->ToString())));
+      std::wstring envValue(path_util::to_wstring(Nan::Utf8String(envValues->Get(i))));
       envBlock << envValue << L'\0';
     }
 
@@ -165,9 +165,9 @@ static NAN_METHOD(PtyStartProcess) {
     goto cleanup;
   }
 
-  int cols = info[4]->Int32Value();
-  int rows = info[5]->Int32Value();
-  bool debug = info[6]->ToBoolean()->IsTrue();
+  int cols = info[4]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  int rows = info[5]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  bool debug = info[6]->ToBoolean(Nan::GetCurrentContext()).ToLocalChecked()->IsTrue();
 
   // Enable/disable debugging
   SetEnvironmentVariable(WINPTY_DBG_VARIABLE, debug ? "1" : NULL); // NULL = deletes variable
@@ -252,9 +252,9 @@ static NAN_METHOD(PtyResize) {
     return;
   }
 
-  int handle = info[0]->Int32Value();
-  int cols = info[1]->Int32Value();
-  int rows = info[2]->Int32Value();
+  int handle = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  int cols = info[1]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  int rows = info[2]->Int32Value(Nan::GetCurrentContext()).FromJust();
 
   winpty_t *pc = get_pipe_handle(handle);
 
@@ -281,8 +281,8 @@ static NAN_METHOD(PtyKill) {
     return;
   }
 
-  int handle = info[0]->Int32Value();
-  HANDLE innerPidHandle = (HANDLE)info[1]->Int32Value();
+  int handle = info[0]->Int32Value(Nan::GetCurrentContext()).FromJust();
+  HANDLE innerPidHandle = (HANDLE)info[1]->Int32Value(Nan::GetCurrentContext()).FromJust();
 
   winpty_t *pc = get_pipe_handle(handle);
   if (pc == nullptr) {
@@ -300,7 +300,7 @@ static NAN_METHOD(PtyKill) {
 * Init
 */
 
-extern "C" void init(v8::Handle<v8::Object> target) {
+extern "C" void init(v8::Local<v8::Object> target) {
   Nan::HandleScope scope;
   Nan::SetMethod(target, "startProcess", PtyStartProcess);
   Nan::SetMethod(target, "resize", PtyResize);

--- a/src/windowsPtyAgent.ts
+++ b/src/windowsPtyAgent.ts
@@ -115,10 +115,8 @@ export class WindowsPtyAgent {
     // TODO: Wait for ready event?
 
     if (this._useConpty) {
-      const connect = (this._ptyNative as IConptyNative).connect(this._pty, commandLine, cwd, env, c => {
-          console.warn('callback', c)
-          this._$onProcessExit(c)
-      });
+      const connect = (this._ptyNative as IConptyNative).connect(this._pty, commandLine, cwd, env, c => this._$onProcessExit(c)
+);
       this._innerPid = connect.pid;
     }
   }

--- a/src/windowsPtyAgent.ts
+++ b/src/windowsPtyAgent.ts
@@ -115,7 +115,10 @@ export class WindowsPtyAgent {
     // TODO: Wait for ready event?
 
     if (this._useConpty) {
-      const connect = (this._ptyNative as IConptyNative).connect(this._pty, commandLine, cwd, env, this._$onProcessExit.bind(this));
+      const connect = (this._ptyNative as IConptyNative).connect(this._pty, commandLine, cwd, env, c => {
+          console.warn('callback', c)
+          this._$onProcessExit(c)
+      });
       this._innerPid = connect.pid;
     }
   }

